### PR TITLE
Stats canvas

### PR DESCRIPTION
**Pull request by Nuclear-Squid**
The bar charts in the stats page weren’t very precise, so I made a new component to split the different bars in up to three different qualities : `good`, `meh` and `bad` (rendered in blue, yellow and orange respectively).

### CODE REVIEW

File: code/layout-analyzer.js [hunk 1]
```diff
@@ -68,6 +70,49 @@ window.addEventListener('DOMContentLoaded', () => {
...
+      : document;
+    element.querySelector(sel).innerText =
+      nums.map(value => fmtPercent(value, precision)).join(' / ');
+  };
+
+  const sumUpBar = bar => bar.good + bar.meh + bar.bad;
+  const sumUpBarGroup = group => group.reduce((acc, bar) => acc + sumUpBar(bar), 0);
+
+  const getKeyPositionQuality = keyCode => {
+    // This is cursed, there *has* to be a better way
```
Comment (fabi1cazenave) [inline code/layout-analyzer.js:85]: Agreed. Defining two key sets (`good`, `meh`) and checking whether `keyCode` belongs to one of them would be smarter.

File: code/stats-canvas.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      table.innerHTML = renderTableRow(
+        values.map(group => group.map(bar => fmtPercent(sumUpBar(bar), precision)))
+      );
+    }
+
+    // const tableEntries = values.map(group =>
+    //   group.map(column => `<td>${totalLoad(column)}</td>`).join('')
+    // ).join(`<td id="empty"></td>`);
+
+    // table.innerHTML = `<tr>${tableEntries}</tr>`;
```
Comment (fabi1cazenave) [inline code/stats-canvas.js:145]: Nit, let’s drop the disabled code.

File: layouts/hypergol.json [hunk 3]
```diff
@@ -0,0 +1,59 @@
+{
```
Comment (fabi1cazenave) [inline layouts/hypergol.json:1]: All files related to Hypergol should probably come with another PR — along with a web page explaining what the layout is about. WDYT ?

File: code/stats-canvas.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    if (detailedValues) {
+      const extractQuality = (values, quality) =>
+        values.map(group => group.map(bar => bar[quality]));
+
+      const notAllZeros = row => row.some(group => group.some(item => item != 0));
+
+      const qualities = Object.keys(this.colors);
+      if (!flipVertically) qualities.reverse();
+
+      // console.log(
```
Comment (fabi1cazenave) [inline code/stats-canvas.js:125]: spotted ! :-)

File: www/layouts/shortcodes/layout-list.html [hunk 5]
```diff
@@ -19,6 +19,7 @@
     "colemak-french-touch" "Colemak French Touch"
     "bepolar"     "Bépolar"
     "bepo"        "Bépo"
+    "hypergol"    "Hypergol"
```
Comment (fabi1cazenave) [inline www/layouts/shortcodes/layout-list.html:22]: Same here : unless I’m missing a point, I think this should come up with another PR.

**Review by fabi1cazenave — COMMENTED** [review]
Looks OK to me. !’ve left a couple suggestions.

As the code complexity grows mostly because of the bar charts, in the future I think we should give D3.js a try (or any similar lib) rather than re-inventing the wheel.

File: code/layout-analyzer.js (hunk 1 above)
Comment (fabi1cazenave) [reply code/layout-analyzer.js:85]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

File: layouts/hypergol.json (hunk 3 above)
Comment (Nuclear-Squid) [reply layouts/hypergol.json:1]: Oops, at first the PR was dedicated to Hypergol, I just needed “a few changes” to make it work. Forgot to remove hypergol once it became clear there was a *lot* of work needed

File: code/layout-analyzer.js (hunk 1 above)
Comment (Nuclear-Squid) [reply code/layout-analyzer.js:85]: I almost never use sets, completely forgot they even existed lol

File: code/layout-analyzer.js [hunk 6]
```diff
@@ -81,36 +81,19 @@ window.addEventListener('DOMContentLoaded', () => {
   const sumUpBar = bar => bar.good + bar.meh + bar.bad;
   const sumUpBarGroup = group => group.reduce((acc, bar) => acc + sumUpBar(bar), 0);
 
+  // This has to be the *stupidest* way to format code, and I love it
+  const goodKeysSet = new Set([
+            'KeyW', 'KeyE',                    'KeyI', 'KeyO',
+    'KeyA', 'KeyS', 'KeyD', 'KeyF',    'KeyJ', 'KeyK', 'KeyL', 'Semicolon',
+                            'KeyV',    'KeyM',
+  ]);
```
Comment (fabi1cazenave) [inline code/layout-analyzer.js:89]: ! love it. Great idea. :-)

**Review by fabi1cazenave — APPROVED** [review]
LGTM

**Pull request merged**